### PR TITLE
Subcell: NewtonianEuler mutators for computing ghost cell data

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Add mutator to compute ghost cell data when doing DG
- Add mutator to compute ghost cell data when doing FD

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/Systems/NewtonianEuler/Subcell/PrimitiveGhostData.hpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#include <cstddef>
+
+#include "DataStructures/VariablesTag.hpp"
+#include "Evolution/Systems/NewtonianEuler/Tags.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace NewtonianEuler::subcell {
+/*!
+ * \brief Computes the mass density, velocity, and pressure on the subcells so
+ * they can be used for reconstruction.
```
Comment (fmahebert) [inline src/Evolution/Systems/NewtonianEuler/Subcell/PrimitiveGhostData.hpp]: I'm a bit confused by what's going on here. Things I think would help:
- remove the "Ghost" from the class name or explain what's "ghostly" about it in the documentation
- clarify the sentence below (lines 17-18) as to which grid the copy is happening on. Is this subcell grid -> subcell grid?

File: tests/Unit/Evolution/Systems/NewtonianEuler/Subcell/Test_PrimitiveGhostData.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  using prim_tags =
+      tmpl::list<MassDensity, Velocity, SpecificInternalEnergy, Pressure>;
+  using PrimVars = Variables<prim_tags>;
+  using prims_to_reconstruct_tags = tmpl::list<MassDensity, Velocity, Pressure>;
+  using ReconsPrimVars = Variables<prims_to_reconstruct_tags>;
+
+  const Mesh<Dim> subcell_mesh{9, Spectral::Basis::FiniteDifference,
+                               Spectral::Quadrature::CellCentered};
+  // NOLINTNEXTLINE(modernize-use-auto)
+  const PrimVars prims = make_with_random_values<PrimVars>(
```
Comment (fmahebert) [inline tests/Unit/Evolution/Systems/NewtonianEuler/Subcell/Test_PrimitiveGhostData.cpp:44]: If I correctly understood things, this would be more clearly named `expected_prims`?

File: tests/Unit/Evolution/Systems/NewtonianEuler/Subcell/Test_PrimitiveGhostData.cpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      gen, dist, subcell_mesh.number_of_grid_points());
+
+  auto box = db::create<db::AddSimpleTags<::Tags::Variables<prim_tags>>>(prims);
+  const ReconsPrimVars recons_prims = db::mutate_apply<
```
Comment (fmahebert) [inline tests/Unit/Evolution/Systems/NewtonianEuler/Subcell/Test_PrimitiveGhostData.cpp]: I think `prims_to_reconstruct` would be clearer to me because it says in the name you haven't yet reconstructed anything (i.e., clarifies this is inputs to a future reconstruction)

File: src/Evolution/Systems/NewtonianEuler/Subcell/PrimitiveGhostData.hpp [hunk 4]
```diff
@@ -41,4 +48,44 @@ class PrimitiveGhostDataOnSubcells {
   static Variables<prims_to_reconstruct_tags> apply(
       const Variables<prim_tags>& prims) noexcept;
 };
+
+/*!
+ * \brief Projects the mass density, velocity, and pressure to the subcells so
+ * they can be used for reconstruction.
```
Comment (fmahebert) [inline src/Evolution/Systems/NewtonianEuler/Subcell/PrimitiveGhostData.hpp]: - As in the previous commit, can you clarify (or remove) the "Ghost" part of the name here? I'm still confused.
- Can you now also clarify the "ToSlice" part of the name? This just talks about projections and reconstructions, and I don't understand what will be sliced where.

File: src/Evolution/Systems/NewtonianEuler/Subcell/PrimitiveGhostData.hpp [hunk 5]
```diff
@@ -41,4 +48,44 @@ class PrimitiveGhostDataOnSubcells {
...
+ *
+ * The computation just copies the data from the primitive variables tag to a
+ * new Variables, then projects that Variables to the subcells. In the future
+ * we will likely want to elide this copy but that requires support from the
+ * actions.
+ *
+ * This mutator is passed what `Metavars::SubcellOptions::GhostDataToSlice` must
+ * be set to.
```
Comment (fmahebert) [inline src/Evolution/Systems/NewtonianEuler/Subcell/PrimitiveGhostData.hpp]: My brain can't parse this... help? maybe rephrase?

File: tests/Unit/Evolution/Systems/NewtonianEuler/Subcell/Test_PrimitiveGhostData.cpp [hunk 6]
```diff
@@ -50,6 +54,45 @@ void test_subcells(
...
+  const ReconsPrimVars recons_prims =
+      db::mutate_apply<NewtonianEuler::subcell::PrimitiveGhostDataToSlice<Dim>>(
+          make_not_null(&box));
+
+  const PrimVars subcell_prims = evolution::dg::subcell::fd::project(
+      prims, dg_mesh, subcell_mesh.extents());
+  tmpl::for_each<prims_to_reconstruct_tags>(
+      [&recons_prims, &subcell_prims](auto tag_v) {
+        using tag = tmpl::type_from<decltype(tag_v)>;
+        CHECK_ITERABLE_APPROX(get<tag>(recons_prims), get<tag>(subcell_prims));
```
Comment (fmahebert) [inline tests/Unit/Evolution/Systems/NewtonianEuler/Subcell/Test_PrimitiveGhostData.cpp]: I don't know what's compared to what... can one of these be prefixed with `expected_`?

As in the other commit, would `recons_prims` -> `prims_to_reconstruct` be accurate? if so, I'd find it easier to put in context.

**Review by fmahebert — COMMENTED** [review]
The code looks okay, and I understand what it is doing, but I have no idea why... so a few clarification requests 😛

File: src/Evolution/Systems/NewtonianEuler/Subcell/PrimitiveGhostData.hpp (hunk 1 above)
Comment (nilsdeppe) [reply src/Evolution/Systems/NewtonianEuler/Subcell/PrimitiveGhostData.hpp]: - this sends what the neighbor would consider "ghost cells" so ghost is the appropriate name. Do you have a suggestion for what to add to the documentation? I'm not sure without going into a full-blown explanation of how FD/FV works, which seems a bit overkill here.

File: tests/Unit/Evolution/Systems/NewtonianEuler/Subcell/Test_PrimitiveGhostData.cpp (hunk 2 above)
Comment (nilsdeppe) [reply tests/Unit/Evolution/Systems/NewtonianEuler/Subcell/Test_PrimitiveGhostData.cpp:44]: sort of? It's also the input prims. It's just that the input and output prims are equal in this case

File: src/Evolution/Systems/NewtonianEuler/Subcell/PrimitiveGhostData.hpp (hunk 1 above)
Comment (fmahebert) [reply src/Evolution/Systems/NewtonianEuler/Subcell/PrimitiveGhostData.hpp]: How about changing "so they can be used for reconstruction" -> "so they can be sent to the neighbors for their reconstructions". For me, once you state clearly this will be sent to a neighbor, I have more context for understanding "ghost".

File: tests/Unit/Evolution/Systems/NewtonianEuler/Subcell/Test_PrimitiveGhostData.cpp (hunk 2 above)
Comment (fmahebert) [reply tests/Unit/Evolution/Systems/NewtonianEuler/Subcell/Test_PrimitiveGhostData.cpp:44]: Agreed, but even so, naming it "expected" will make it easier to understand in the CHECK statement what's being tested against what. I don't feel strongly about this, so do as you prefer :)

**Comment by nilsdeppe** [comment]
Thanks for the quick responses to my questions! I've pushed fixups :)

**Review by fmahebert — COMMENTED** [review]
LGTM, please squash

**Comment by nilsdeppe** [comment]
Okay, squashed. Thanks :)

File: src/Evolution/Systems/NewtonianEuler/Subcell/PrimitiveGhostData.hpp [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+  using prim_tags =
+      tmpl::list<MassDensity, Velocity, SpecificInternalEnergy, Pressure>;
+  using prims_to_reconstruct_tags = tmpl::list<MassDensity, Velocity, Pressure>;
+
+ public:
+  using return_tags = tmpl::list<>;
+  using argument_tags = tmpl::list<::Tags::Variables<prim_tags>>;
+
+  static Variables<prims_to_reconstruct_tags> apply(
```
Comment (kidder) [inline src/Evolution/Systems/NewtonianEuler/Subcell/PrimitiveGhostData.hpp:51]: forward declare Variables

File: src/Evolution/Systems/NewtonianEuler/Subcell/PrimitiveGhostData.cpp [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+#include "DataStructures/VariablesTag.hpp"
+#include "Evolution/Systems/NewtonianEuler/Tags.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace NewtonianEuler::subcell {
+template <size_t Dim>
+auto PrimitiveGhostDataOnSubcells<Dim>::apply(
+    const Variables<prim_tags>& prims) noexcept
```
Comment (kidder) [inline src/Evolution/Systems/NewtonianEuler/Subcell/PrimitiveGhostData.cpp:19]: include Variables

File: src/Evolution/Systems/NewtonianEuler/Subcell/PrimitiveGhostData.hpp [hunk 9]
```diff
@@ -42,4 +49,48 @@ class PrimitiveGhostDataOnSubcells {
   static Variables<prims_to_reconstruct_tags> apply(
       const Variables<prim_tags>& prims) noexcept;
 };
+
+/*!
+ * \brief Projects the mass density, velocity, and pressure to the subcells so
+ * they can be sent to the neighbors for their reconstructions.
+ *
+ * The data on the subcells is sliced before sending to the neighbors. That is,
```
Comment (kidder) [inline src/Evolution/Systems/NewtonianEuler/Subcell/PrimitiveGhostData.hpp]: I was confused by this as it does not appear to me that the slicing is done in this function

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash

**Comment by nilsdeppe** [comment]
I've rebased and squashed in the fixes. Thanks for the reviews :)

**Review by fmahebert — DISMISSED** [review]

**Review by kidder — DISMISSED** [review]

**Comment by kidder** [comment]
needs a rebase

**Comment by nilsdeppe** [comment]
I've rebased this on top of #3225 so that hopefully it won't need a second rebase :crossed_fingers:

**Review by kidder — APPROVED** [review]

**Review by fmahebert — APPROVED** [review]
